### PR TITLE
Connector for Google Drive

### PR DESCRIPTION
**Pull request by HAKSOAT**
Implements enhancement #244

### CODE REVIEW

**Comment by HAKSOAT** [comment]
Hi, @cragwolfe I'd appreciate it if a review could begin on my PR.

Everything works as expected from running it, but I am looking to do two more things:

Integrate Google's Auth ID (so files can be downloaded without making them available for everyone to view) and find a way to allow downloads on public files without a token.

I'm not sure why the CI CD checks are not running at the moment.

No tests at the moment since API keys are needed (I'm not sure how to go about creating a test for that without exposing keys).

**Comment by cragwolfe** [comment]
> Hi, @cragwolfe I'd appreciate it if a review could begin on my PR.
> 
Hi @HAKSOAT , happy to! Great to hear you got this working.

> Everything works as expected from running it, but I am looking to do two more things:
> 
> Integrate Google's Auth ID (so files can be downloaded without making them available for everyone to view) and find a way to allow downloads on public files without a token.
> 
Does that mean following the `credentials.json` flow described here: https://developers.google.com/drive/api/quickstart/python ?

> I'm not sure why the CI CD checks are not running at the moment.
> 
I think it was mostly the gihub outage. I just merged into Main to help trigger that (hope that's OK!).

> No tests at the moment since API keys are needed (I'm not sure how to go about creating a test for that without exposing keys).

How does one go about getting an API Key?

A few more things to address:

* Even though it needs to be verified manually, please add a test script under https://github.com/Unstructured-IO/unstructured/tree/main/examples/ingest/google-drive/ . The script itself can provide guidance on how to get authentication credentials.

* Please update `CHANGELOG.md` and `unstructured/__version__.py`.

* Running `make tidy` and `make check` should help you past the CI lint issue.

File: unstructured/ingest/connector/google_drive.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+                    else:
+                        ext = ""
+                        if not Path(meta.get("name")).suffixes:
+                            guess = guess_extension(meta.get("mimeType"))
+                            ext = guess if guess else ext
+
+                        if meta.get("mimeType", "").startswith("application/vnd.google-apps"):
+                            export_mime = GOOGLE_DRIVE_EXPORT_TYPES.get(meta.get("mimeType"))
+                            if not export_mime:
+                                continue
```
Comment (cragwolfe) [inline unstructured/ingest/connector/google_drive.py:161]: it be nice to print what we are skipping here

File: unstructured/ingest/connector/google_drive.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+                            if not ext:
+                                guess = guess_extension(export_mime)
+                                ext = guess if guess else ext
+
+                        # TODO (Habeeb): Consider filtering at the query level.
+                        if self.config.extension and self.config.extension != ext:
+                            continue
```
Comment (cragwolfe) [inline unstructured/ingest/connector/google_drive.py:169]: it be nice to print what we are skipping here (again)

Comment (cragwolfe) [reply unstructured/ingest/connector/google_drive.py:169]: E.g., https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/ingest/connector/github.py does this nicely: both prints what is skipped and total number of files skipped.

**Comment by HAKSOAT** [comment]
Hi @cragwolfe 

Thanks for the review. I'll go through your comments now.

> Does that mean following the `credentials.json` flow described here: https://developers.google.com/drive/api/quickstart/python ?
> 

More like here: 

https://developers.google.com/workspace/guides/create-credentials

I believe a service account is better as it is made for robot accounts. The reason is that the regular API key gets throttled as I experienced an error after a couple of requests that says:

> <h1>We're sorry...</h1><p>... but your computer or network may be sending automated queries. To protect our users, we can't process your request right now.</p></div><div style="margin-left: 4em;">See <a href="https://support.google.com/websearch/answer/86640">Google Help</a> for more information.

**Comment by HAKSOAT** [comment]
Hi @cragwolfe 

I think I just hit a roadblock for service accounts. Looks like there's an issue with using the service account approach with multiprocessing.


```  
Traceback (most recent call last):
  File "/mnt/c/Users/HAKS/dev/unstructured/./unstructured/ingest/main.py", line 362, in <module>
    main()
  File "/home/haks/Envs/unstructured/lib/python3.9/site-packages/click/core.py", line 1126, in __call__
    return self.main(*args, **kwargs)
  File "/home/haks/Envs/unstructured/lib/python3.9/site-packages/click/core.py", line 1051, in main
    rv = self.invoke(ctx)
  File "/home/haks/Envs/unstructured/lib/python3.9/site-packages/click/core.py", line 1393, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/haks/Envs/unstructured/lib/python3.9/site-packages/click/core.py", line 752, in invoke
    return __callback(*args, **kwargs)
  File "/mnt/c/Users/HAKS/dev/unstructured/./unstructured/ingest/main.py", line 353, in main
    MainProcess(
  File "/mnt/c/Users/HAKS/dev/unstructured/./unstructured/ingest/main.py", line 73, in run
    results = pool.map(self.doc_processor_fn, docs)  # noqa: F841
  File "/home/haks/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/haks/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/home/haks/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/pool.py", line 537, in _handle_tasks
    put(task)
  File "/home/haks/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/connection.py", line 211, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/haks/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_cffi_backend.FFI' object
```

The traceback isn't so helpful either. Just putting this here to serve as an update as I have tried to find a fix but I am yet to find one. I will take another look in the morning here.

**Comment by cragwolfe** [comment]
> I think I just hit a roadblock for service accounts. Looks like there's an issue with using the service account approach with multiprocessing.

Yes, looks that way. A way around that is to remove `self.service` from the `config` object, and only create the service object on demand (it looks like the `api_key`, and `drive_id` are sufficient to build the service object).

**Comment by HAKSOAT** [comment]
> Yes, looks that way. A way around that is to remove `self.service` from the `config` object, and only create the service object on demand (it looks like the `api_key`, and `drive_id` are sufficient to build the service object).

Thanks for the hint @cragwolfe It works now. I think a second review will be great so I can fix any other issue.

**Comment by HAKSOAT** [comment]
Hi @cragwolfe please can you help take a look at this when chanced?

**Comment by cragwolfe** [comment]
> Hi @cragwolfe please can you help take a look at this when chanced?
For sure, on my list for today!

**Review by cragwolfe — APPROVED** [review]
Works very nicely! 🚀 

Though, I did get tripped up by not reading the Note in ingest.sh carefully enough the first time:

```
# NOTE: Using the Service Account key only works when the file or folder
# is shared atleast with permission for "Anyone with the link" to view
# OR the email address for the service account is given access to the file
# or folder.
```
service account emails **do not** look like user emails.

**Pull request merged**